pfcon/pfcon.py

```diff
@@ -1973,3 +1959,79 @@ def setup(self, **kwargs):
             Colors.NO_COLOUR,
             level   = 1,
             syslog  = False)
+
+    def get_pfioh_ip(self):
+        pfioh_ip = (
+            self.read_ip_by_host_name_environment_variable(
+                PFIOH_HOST_NAME_ENVIRONMENT_VARIABLE,
+                default_host_name=PFIOH_HOST_NAME_DEFAULT)
+            or self.read_from_environment(PFIOH_IP_ENVIRONMENT_VARIABLE)
+            or self.read_from_environment(HOST_IP_ENVIRONMENT_VARIABLE)
+            or read_local_host_ip()
+        )
+        self.dp.qprint(f'Found pfioh IP address: {pfioh_ip}.')
+        return pfioh_ip
+
+    def get_pman_ip(self):
+        pman_ip = (
+            self.read_ip_by_host_name_environment_variable(
+                PMAN_HOST_NAME_ENVIRONMENT_VARIABLE,
+                default_host_name=PMAN_HOST_NAME_DEFAULT)
+            or self.read_from_environment(PMAN_IP_ENVIRONMENT_VARIABLE)
+            or self.read_from_environment(HOST_IP_ENVIRONMENT_VARIABLE)
+            or read_local_host_ip()
+        )
+        self.dp.qprint(f'Found pman IP address: {pman_ip}.')
+        return pman_ip
+
+    def read_ip_by_host_name_environment_variable(
+            self, host_name_environment_variable, default_host_name=''):
+        '''For newer docker-compose'''
+
+        host_name = os.environ.get(
+            host_name_environment_variable, default_host_name
+        )
+        self.dp.qprint(f'Reading IP address of {host_name}.')
+        try:
+            ip_address = socket.gethostbyname(host_name)
+        except Exception as exception:
+            self.dp.qprint(
+                f'Could not obtain IP address of host {host_name} ({exception}).'
+            )
+            return None
+
+        if ip_address == '127.0.0.1':
+            return None
+        return ip_address
+
+
+    def read_from_environment(self, environment_variable, default=None):
+        value = os.environ.get(environment_variable, default)
+        self.dp.qprint(
+            f'Obtained {value} from environment variable {environment_variable}'
+            f' (default: {default})'
+        )
+        return value
+
+
+def read_local_host_ip():
+    local_host_ip = [
+        l for l in (
+            [
+                ip for ip in socket.gethostbyname_ex(socket.gethostname())[2]
+                if not ip.startswith("127.")
+            ][:1],
+            [
+                [
+                    (
+                        s.connect(('8.8.8.8', 53)),
```

Comment:
I see the IP address `8.8.8.8` for DNS is hardcorded. Is it fine not to be configurable?

Reply:
That was part of the original code. I just reformatted it for clarity. To be honest, I don't know what this piece of code is doing exactly...

Reply:
It's a contrived way to get the machine's local IP address, that's it. @rudolphpienaar probably put it in for convenience.

https://github.com/FNNDSC/pfcon/pull/76/files/065189c5531c5e530b37d39c0a62f38c5f1a2d9c#diff-bedea3f9ced94f40f2c9b6cf2cd7c9ed25d3f515036a3f3d2817ffe0600f6c41L1882

The line is found repeated in many of our repos:

- https://github.com/FNNDSC/pfurl/blob/f355ef18d59b153d0ef03915b02ba241a17aefac/bin/pfurl#L23
- https://github.com/FNNDSC/pman/blob/d65f9aa500924c7c1fb7cb965c546fd000ae4c87/bin/pman#L31

IMO the entire functionality of reading the private IP address is bloat. If we want to keep it for convenience, there are shorter ways of figuring it out. Though less adaptable -- that socket connection to 8.8.8.8 is a fallback workaround in case a previous try to find out the IP did not work in an esoteric environment. See https://stackoverflow.com/questions/166506/finding-local-ip-addresses-using-pythons-stdlib/166589#166589

I was scratching my head about this a few months ago too, if you want a thorough explanation of what's happening please see [pfmisc/local_ip.py](https://github.com/FNNDSC/pfmisc/blob/master/pfmisc/local_ip.py) where I flattened the control flow and commented.

Reply:
Great, thanks for the insights @jennydaman :+1:

Reply:
> That was part of the original code.

Okay. @mamurak @jennydaman thanks for sharing the context and documents!
It seems the 8.8.8.8 is Google's public DNS server IP.
https://developers.google.com/speed/public-dns/

Reply:
Yes, I am the source of these convolutions. The reason it was necessary arose the proxied env of our Institution network. The short story is that addresses such as `localhost` were not handled consistently in the proxied env and hence I needed to code/borrow some quick method to find the actual host IP address (and not `localhost` or `127.0.0.1`).

Yes, 8.8.8.8 is one of google's public DNS -- it has been that way for many years so I assume that IP is reliable/stable.